<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Getting Started • rsynthbio</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Getting Started">
<link rel="icon" type="image/png" href="https://github.com/synthesizebio/rsynthbio/tree/main/assets/logomark.png?raw=true">
<link rel="shortcut icon" type="image/png" href="https://github.com/synthesizebio/rsynthbio/tree/main/assets/logomark.png?raw=true">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">
    <img src="https://github.com/synthesizebio/rsynthbio/blob/main/assets/whitelogo.png?raw=true" width="200px"><a class="navbar-brand me-2" href="../index.html">rsynthbio</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">4.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/getting-started.html">Getting Started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/synthesizebio/rsynthbio" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.synthesize.bio/">Synthesize.bio main page</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting Started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/synthesizebio/rsynthbio/blob/main/vignettes/getting-started.Rmd" class="external-link"><code>vignettes/getting-started.Rmd</code></a></small>
      <div class="d-none name"><code>getting-started.Rmd</code></div>
    </div>

    
    
<p><code>rsynthbio</code> is an R package that provides a convenient
interface to the <a href="https://www.synthesize.bio/" class="external-link">Synthesize
Bio</a> API, allowing users to generate realistic gene expression data
based on specified biological conditions. This package enables
researchers to easily access AI-generated transcriptomic data for
various modalities including bulk RNA-seq and single-cell RNA-seq.</p>
<p>Alternatively, you can AI generate datasets from our <a href="https://app.synthesize.bio/datasets/" class="external-link">web platform</a>.</p>
<div class="section level2">
<h2 id="how-to-install">How to install<a class="anchor" aria-label="anchor" href="#how-to-install"></a>
</h2>
<p>You can install <code>rsynthbio</code> from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rsynthbio"</span><span class="op">)</span></span></code></pre></div>
<p>If you want the development version, you can install using the
<code>remotes</code> package to install from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="st">"remotes"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"synthesizebio/rsynthbio"</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, load the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/synthesizebio/rsynthbio" class="external-link">rsynthbio</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="authentication">Authentication<a class="anchor" aria-label="anchor" href="#authentication"></a>
</h2>
<p>Before using the Synthesize Bio API, you need to set up your API
token. The package provides a secure way to handle authentication:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Securely prompt for and store your API token</span></span>
<span><span class="co"># The token will not be visible in the console</span></span>
<span><span class="fu"><a href="../reference/set_synthesize_token.html">set_synthesize_token</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># You can also store the token in your system keyring for persistence</span></span>
<span><span class="co"># across R sessions (requires the 'keyring' package)</span></span>
<span><span class="fu"><a href="../reference/set_synthesize_token.html">set_synthesize_token</a></span><span class="op">(</span>use_keyring <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Loading your API key for a session.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># In future sessions, load the stored token</span></span>
<span><span class="fu"><a href="../reference/load_synthesize_token_from_keyring.html">load_synthesize_token_from_keyring</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check if a token is already set</span></span>
<span><span class="fu"><a href="../reference/has_synthesize_token.html">has_synthesize_token</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can obtain an API token by registering at <a href="https://app.synthesize.bio" class="external-link">Synthesize Bio</a>.</p>
<div class="section level3">
<h3 id="security-best-practices">Security Best Practices<a class="anchor" aria-label="anchor" href="#security-best-practices"></a>
</h3>
<p>For security reasons, remember to clear your token when you’re
done:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clear token from current session</span></span>
<span><span class="fu"><a href="../reference/clear_synthesize_token.html">clear_synthesize_token</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clear token from both session and keyring</span></span>
<span><span class="fu"><a href="../reference/clear_synthesize_token.html">clear_synthesize_token</a></span><span class="op">(</span>remove_from_keyring <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Never hard-code your token in scripts that will be shared or
committed to version control.</p>
</div>
</div>
<div class="section level2">
<h2 id="designing-queries-for-models">Designing Queries for Models<a class="anchor" aria-label="anchor" href="#designing-queries-for-models"></a>
</h2>
<div class="section level3">
<h3 id="choosing-a-model">Choosing a Model<a class="anchor" aria-label="anchor" href="#choosing-a-model"></a>
</h3>
<p>The first step is to identify which model you want to use for
prediction:</p>
<ul>
<li>
<strong><code>gem-1-bulk</code></strong>: Bulk RNA-seq (asynchronous
under the hood, returned as data frames)</li>
<li>
<strong><code>gem-1-sc</code></strong>: Single-cell RNA-seq
(asynchronous under the hood, returned as data frames)</li>
</ul>
<p>You can check which models are available programmatically:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check available models</span></span>
<span><span class="fu"><a href="../reference/list_models.html">list_models</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a query for the bulk model</span></span>
<span><span class="va">bulk_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">bulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_query.html">predict_query</a></span><span class="op">(</span><span class="va">bulk_query</span>, model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a query for the single-cell model</span></span>
<span><span class="va">sc_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-sc"</span><span class="op">)</span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_query.html">predict_query</a></span><span class="op">(</span><span class="va">sc_query</span>, model_id <span class="op">=</span> <span class="st">"gem-1-sc"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="creating-a-query">Creating a Query<a class="anchor" aria-label="anchor" href="#creating-a-query"></a>
</h3>
<p>The structure of the query required by the API is specific to each
model. You can use <code><a href="../reference/get_example_query.html">get_example_query()</a></code> to get a correctly
structured example for your chosen model.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the example query structure for a specific model</span></span>
<span><span class="va">example_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the query structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">example_query</span><span class="op">)</span></span></code></pre></div>
<p>The query consists of:</p>
<ol style="list-style-type: decimal">
<li>
<strong><code>mode</code></strong>: The prediction mode that
controls how expression data is generated:
<ul>
<li>
<strong>“sample generation”</strong>: Generates realistic-looking
synthetic data with measurement error (bulk only)</li>
<li>
<strong>“mean estimation”</strong>: Provides stable mean estimates
of expression levels (bulk and single-cell)</li>
</ul>
</li>
<li>
<strong><code>inputs</code></strong>: A list of biological
conditions to generate data for</li>
</ol>
<p>Each input contains <code>metadata</code> (describing the biological
sample) and <code>num_samples</code> (how many samples to generate).</p>
<blockquote>
<p>See the <a href="#query-parameters">Query Parameters</a> section
below for detailed documentation on <code>mode</code> and other optional
query fields.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="making-a-prediction">Making a Prediction<a class="anchor" aria-label="anchor" href="#making-a-prediction"></a>
</h3>
<p>Once your query is ready, you can send it to the API to generate gene
expression data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_query.html">predict_query</a></span><span class="op">(</span><span class="va">query</span>, model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span></code></pre></div>
<p>This result will be a list of two dataframes: <code>metadata</code>
and <code>expression</code></p>
</div>
<div class="section level3">
<h3 id="understanding-the-async-api">Understanding the Async API<a class="anchor" aria-label="anchor" href="#understanding-the-async-api"></a>
</h3>
<p>Behind the scenes, the API uses an <strong>asynchronous
model</strong> to handle queries efficiently:</p>
<ol style="list-style-type: decimal">
<li>Your query is submitted to the API, which returns a query ID</li>
<li>The function automatically polls the status endpoint (default: every
2 seconds)</li>
<li>When the query completes, results are downloaded from a signed
URL</li>
<li>Data is parsed and returned as R data frames</li>
</ol>
<p>All of this happens automatically when you call
<code><a href="../reference/predict_query.html">predict_query()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="controlling-async-behavior">Controlling Async Behavior<a class="anchor" aria-label="anchor" href="#controlling-async-behavior"></a>
</h3>
<p>You can customize the polling behavior if needed:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Increase timeout for large queries (default: 900 seconds = 15 minutes)</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_query.html">predict_query</a></span><span class="op">(</span></span>
<span>  <span class="va">query</span>,</span>
<span>  model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span>,</span>
<span>  poll_timeout_seconds <span class="op">=</span> <span class="fl">1800</span>, <span class="co"># 30 minutes</span></span>
<span>  poll_interval_seconds <span class="op">=</span> <span class="fl">5</span> <span class="co"># Check every 5 seconds instead of 2</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="valid-metadata-keys">Valid Metadata Keys<a class="anchor" aria-label="anchor" href="#valid-metadata-keys"></a>
</h3>
<p>The input metadata is a list of lists. This is the full list of valid
metadata keys:</p>
<p><strong>Biological:</strong></p>
<ul>
<li><code>age_years</code></li>
<li><code>cell_line_ontology_id</code></li>
<li><code>cell_type_ontology_id</code></li>
<li><code>developmental_stage</code></li>
<li><code>disease_ontology_id</code></li>
<li><code>ethnicity</code></li>
<li><code>genotype</code></li>
<li><code>race</code></li>
<li>
<code>sample_type</code> (“cell line”, “organoid”, “other”, “primary
cells”, “primary tissue”, “xenograft”)</li>
<li>
<code>sex</code> (“male”, “female”)</li>
<li><code>tissue_ontology_id</code></li>
</ul>
<p><strong>Perturbational:</strong></p>
<ul>
<li><code>perturbation_dose</code></li>
<li><code>perturbation_ontology_id</code></li>
<li><code>perturbation_time</code></li>
<li>
<code>perturbation_type</code>
(“coculture”,“compound”,“control”,“crispr”,“genetic”,“infection”,“other”,“overexpression”,“peptide
or biologic”,“shrna”,“sirna”)</li>
</ul>
<p><strong>Technical:</strong></p>
<ul>
<li>
<code>study</code> (Bioproject ID)</li>
<li>
<code>library_selection</code> (e.g., “cDNA”, “polyA”, “Oligo-dT” -
see <a href="https://ena-docs.readthedocs.io/en/latest/submit/reads/webin-cli.html#permitted-values-for-library-selection" class="external-link uri">https://ena-docs.readthedocs.io/en/latest/submit/reads/webin-cli.html#permitted-values-for-library-selection</a>)</li>
<li>
<code>library_layout</code> (“PAIRED”, “SINGLE”)</li>
<li>
<code>platform</code> (“illumina”)</li>
</ul>
</div>
<div class="section level3">
<h3 id="valid-metadata-values">Valid Metadata Values<a class="anchor" aria-label="anchor" href="#valid-metadata-values"></a>
</h3>
<p>The following are the valid values or expected formats for selected
metadata keys:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>Metadata Field</th>
<th>Requirement / Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>cell_line_ontology_id</code></td>
<td>Requires a <a href="https://www.cellosaurus.org/" class="external-link">Cellosaurus
ID</a>.</td>
</tr>
<tr class="even">
<td><code>cell_type_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/cl" class="external-link">CL
ID</a>.</td>
</tr>
<tr class="odd">
<td><code>disease_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/mondo" class="external-link">MONDO ID</a>.</td>
</tr>
<tr class="even">
<td><code>perturbation_ontology_id</code></td>
<td>Must be a valid Ensembl gene ID (e.g.,
<code>ENSG00000156127</code>), <a href="https://www.ebi.ac.uk/chebi/" class="external-link">ChEBI ID</a> (e.g.,
<code>CHEBI:16681</code>), <a href="https://www.ebi.ac.uk/chembl/" class="external-link">ChEMBL ID</a> (e.g.,
<code>CHEMBL1234567</code>), or <a href="https://www.ncbi.nlm.nih.gov/taxonomy" class="external-link">NCBI Taxonomy ID</a> (e.g.,
<code>9606</code>).</td>
</tr>
<tr class="odd">
<td><code>tissue_ontology_id</code></td>
<td>Requires a <a href="https://www.ebi.ac.uk/ols4/ontologies/uberon" class="external-link">UBERON ID</a>.</td>
</tr>
</tbody>
</table>
<p>We highly recommend using the <a href="https://www.ebi.ac.uk/ols4/" class="external-link">EMBL-EBI Ontology Lookup Service</a>
to find valid IDs for your metadata.</p>
<p>Models have a limited acceptable range of metadata input values. If
you provide a value that is not in the acceptable range, the API will
return an error.</p>
</div>
<div class="section level3">
<h3 id="query-parameters">Query Parameters<a class="anchor" aria-label="anchor" href="#query-parameters"></a>
</h3>
<p>In addition to metadata, queries support several optional parameters
that control the generation process:</p>
<div class="section level4">
<h4 id="mode-character-required">mode (character, required)<a class="anchor" aria-label="anchor" href="#mode-character-required"></a>
</h4>
<p>Controls the type of prediction the model generates. This parameter
is required in all queries.</p>
<p>Available modes:</p>
<ul>
<li><p><strong>“sample generation”</strong>: The model works identically
to the mean estimation approach, except that the final gene expression
distribution is also sampled to generate realistic-looking synthetic
data that captures the error associated with measurements. This mode is
useful when you want data that mimics real experimental
measurements.</p></li>
<li><p><strong>“mean estimation”</strong>: The model creates a
distribution capturing the biological heterogeneity consistent with the
supplied metadata. This distribution is then sampled to predict a gene
expression distribution that captures measurement error. The mean of
that distribution serves as the prediction. This mode is useful when you
want a stable estimate of expected expression levels.</p></li>
</ul>
<blockquote>
<p><strong>Note:</strong> <strong>Single-cell queries only support “mean
estimation” mode.</strong> Bulk queries support both modes.</p>
</blockquote>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bulk query with sample generation (default for bulk)</span></span>
<span><span class="va">bulk_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">bulk_query</span><span class="op">$</span><span class="va">mode</span> <span class="op">&lt;-</span> <span class="st">"sample generation"</span></span>
<span></span>
<span><span class="co"># Bulk query with mean estimation</span></span>
<span><span class="va">bulk_query_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">bulk_query_mean</span><span class="op">$</span><span class="va">mode</span> <span class="op">&lt;-</span> <span class="st">"mean estimation"</span></span>
<span></span>
<span><span class="co"># Single-cell query (must use mean estimation)</span></span>
<span><span class="va">sc_query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-sc"</span><span class="op">)</span></span>
<span><span class="va">sc_query</span><span class="op">$</span><span class="va">mode</span> <span class="op">&lt;-</span> <span class="st">"mean estimation"</span> <span class="co"># Required for single-cell</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="total_count-integer-optional">total_count (integer, optional)<a class="anchor" aria-label="anchor" href="#total_count-integer-optional"></a>
</h4>
<p>Library size used when converting predicted log CPM back to raw
counts. Higher values scale counts up proportionally.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a query and add custom total_count</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">total_count</span> <span class="op">&lt;-</span> <span class="fl">5000000</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="deterministic_latents-logical-optional">deterministic_latents (logical, optional)<a class="anchor" aria-label="anchor" href="#deterministic_latents-logical-optional"></a>
</h4>
<p>If <code>TRUE</code>, the model uses the mean of each latent
distribution (<code>p(z|metadata)</code> or <code>q(z|x)</code>) instead
of sampling. This removes randomness from latent sampling and produces
deterministic outputs for the same inputs.</p>
<ul>
<li>Default: <code>FALSE</code> (sampling is enabled)</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a query and enable deterministic latents</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">deterministic_latents</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="seed-integer-optional">seed (integer, optional)<a class="anchor" aria-label="anchor" href="#seed-integer-optional"></a>
</h4>
<p>Random seed for reproducibility when using stochastic sampling.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a query with a specific seed</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">42</span></span></code></pre></div>
<p>You can combine multiple parameters in a single query:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a query and add multiple parameters</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">total_count</span> <span class="op">&lt;-</span> <span class="fl">8000000</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">deterministic_latents</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">mode</span> <span class="op">&lt;-</span> <span class="st">"mean estimation"</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_query.html">predict_query</a></span><span class="op">(</span><span class="va">query</span>, model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="modifying-query-inputs">Modifying Query Inputs<a class="anchor" aria-label="anchor" href="#modifying-query-inputs"></a>
</h3>
<p>You can customize the query inputs to fit your specific research
needs:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a base query</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_example_query.html">get_example_query</a></span><span class="op">(</span>model_id <span class="op">=</span> <span class="st">"gem-1-bulk"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adjust number of samples for the first input</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">inputs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">num_samples</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span></span>
<span><span class="co"># Add a new condition</span></span>
<span><span class="va">query</span><span class="op">$</span><span class="va">inputs</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    sex <span class="op">=</span> <span class="st">"male"</span>,</span>
<span>    sample_type <span class="op">=</span> <span class="st">"primary tissue"</span>,</span>
<span>    tissue_ontology_id <span class="op">=</span> <span class="st">"UBERON:0002371"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  num_samples <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="working-with-results">Working with Results<a class="anchor" aria-label="anchor" href="#working-with-results"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Access metadata and expression matrices</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">result</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">result</span><span class="op">$</span><span class="va">expression</span></span>
<span></span>
<span><span class="co"># Check dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">expression</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View metadata sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<p>You may want to process the data in chunks or save it for later
use:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save results to RDS file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">result</span>, <span class="st">"synthesize_results.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load previously saved results</span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"synthesize_results.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Export as CSV</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">expression</span>, <span class="st">"expression_matrix.csv"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">result</span><span class="op">$</span><span class="va">metadata</span>, <span class="st">"sample_metadata.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="additional-resources">Additional Resources<a class="anchor" aria-label="anchor" href="#additional-resources"></a>
</h2>
<ul>
<li><a href="https://github.com/synthesizebio/rsynthbio" class="external-link">Package Source
Code</a></li>
<li><a href="https://github.com/synthesizebio/rsynthbio/issues" class="external-link">File Bug
Reports</a></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Synthesize Bio.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
